src/upload/mod.rs

```diff
@@ -263,6 +265,75 @@ pub async fn upload_package_to_anaconda(
     Ok(())
 }
 
+/// Uploads a package to a channel in an S3 bucket.
+pub async fn upload_package_to_s3(
+    storage: &AuthenticationStorage,
+    channel: Url,
+    endpoint_url: Option<Url>,
+    region: Option<String>,
+    force_path_style: Option<bool>,
+    package_files: &Vec<PathBuf>,
+) -> miette::Result<()> {
+    // If either endpoint_url or region is set, ensure that both are set
+    if endpoint_url.is_some() != region.is_some() {
+        return Err(miette::miette!(
+            "Both endpoint_url and region must be specified together"
+        ));
+    }
+    // TODO: Add CLI args for SECRET_ACCESS_KEY, ACCESS_KEY_ID, and SESSION_TOKEN
+    // and insert custom auth storage backend at position zero
+    let s3_config = if let (Some(endpoint_url), Some(region), Some(force_path_style)) =
+        (endpoint_url, region, force_path_style)
+    {
+        S3Config::Custom {
+            endpoint_url,
+            region,
+            force_path_style,
+        }
+    } else {
+        S3Config::FromAWS
+    };
+
+    let bucket = channel
+        .host_str()
+        .ok_or_else(|| miette::miette!("Failed to get host from channel URL"))?;
+
+    let config_map = HashMap::from([(bucket.to_string(), s3_config.clone())]);
+    let s3 = S3::new(config_map, storage.clone());
+    let s3_client = s3
+        .create_s3_client(channel.clone())
+        .await
+        .map_err(|e| miette::miette!(e.to_string()))?;
```

Comment:
aren't we usually using `.into_diagnostic` for these things?

Reply:
We're returning an `anyhow::Error` from `create_s3_client`, which I believe does not implement [`IntoDiagnostic`](https://docs.rs/miette/latest/miette/trait.IntoDiagnostic.html) (it's not a `std::error::Error`?). Therefore, we have to map the error here.